Mathlib/Data/Polynomial/RingDivision.lean

```diff
@@ -394,6 +394,57 @@ theorem pow_rootMultiplicity_not_dvd {p : R[X]} (p0 : p ≠ 0) (a : R) :
     ¬(X - C a) ^ (rootMultiplicity a p + 1) ∣ p := by rw [← rootMultiplicity_le_iff p0]
 #align polynomial.pow_root_multiplicity_not_dvd Polynomial.pow_rootMultiplicity_not_dvd
 
+theorem Monic.mem_nonZeroDivisors {p : R[X]} (h : p.Monic) : p ∈ nonZeroDivisors R[X] :=
+  mem_nonZeroDivisors_iff.2 fun _ hx ↦ (mul_left_eq_zero_iff h).1 hx
+
+theorem mem_nonZeroDivisors_of_leadingCoeff_mem_nonZeroDivisors {p : R[X]}
+    (h : p.leadingCoeff ∈ nonZeroDivisors R) : p ∈ nonZeroDivisors R[X] := by
+  refine mem_nonZeroDivisors_iff.2 fun x hx ↦ leadingCoeff_eq_zero.1 ?_
+  by_contra hx'
+  rw [← mul_right_mem_nonZeroDivisors_eq_zero_iff h] at hx'
+  simp only [← leadingCoeff_mul' hx', hx, leadingCoeff_zero, not_true] at hx'
+
+theorem rootMultiplicity_mul_X_sub_C_pow' {p : R[X]} {a : R} {n : ℕ} (h : p * (X - C a) ^ n ≠ 0) :
+    (p * (X - C a) ^ n).rootMultiplicity a = p.rootMultiplicity a + n := by
+  refine le_antisymm ?_ ?_
+  · rw [rootMultiplicity_le_iff h]
+    by_contra h'
+    obtain ⟨q, h'⟩ := h'
+    rw [mul_comm p _, add_comm _ n, add_assoc, pow_add, mul_assoc,
+      mul_cancel_left_mem_nonZeroDivisors (monic_X_sub_C a |>.pow n |>.mem_nonZeroDivisors)] at h'
+    exact pow_rootMultiplicity_not_dvd (left_ne_zero_of_mul h) a ⟨q, h'⟩
+  · rw [le_rootMultiplicity_iff h, pow_add]
+    exact mul_dvd_mul_right (pow_rootMultiplicity_dvd p a) _
+
+theorem rootMultiplicity_mul_X_sub_C_pow {p : R[X]} {a : R} {n : ℕ} (h : p ≠ 0) :
```

Comment:
Can you inline the proof of `rootMultiplicity_mul_X_sub_C_pow'` into this one? The condition `p * (X - C a) ^ n ≠ 0` is equivalent to `p ≠ 0` but seems harder to verify.

Reply:
But later I actually need the `rootMultiplicity_mul_X_sub_C_pow'` version to prove things...

Reply:
If you know `p * (X - C a) ^ n ≠ 0`, you can use [left_ne_zero_of_mul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Basic.html#left_ne_zero_of_mul) before applying `rootMultiplicity_mul_X_sub_C_pow`.